[vallox] Vallox Binding initial contribution

SuperOok opened the pull request:
Hi @kaikreuzer ,

I've created a binding for the Vallox central venting unit at https://github.com/SuperOok/openhab-vallox . From there it was successfully tested by me and some other user. 

Now I copied it to the openhab2-addons repository in a branch. I'm not a git-guru so I hope I did everything right... what would be the next steps to get this binding approved eventually?

Cheers,
Hauke

Fixes #1559

SuperOok commented:
How's this PR linked to issue #1559 ?

kubawolanin commented on addons/binding/org.openhab.binding.vallox/ESH-INF/i18n/vallox_xx_XX.properties:
```diff
@@ -0,0 +1,11 @@
+
+# channel types
+channel-type.vallox.sample-channel.label = <Your localized Channel label>
+channel-type.vallox.sample-channel.description = <Your localized Channel description>
```
Please remove this file

kubawolanin commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,413 @@
+				<label>Host</label>
+				<description>Host name or IP address of RS485-to-Ethernet gateway. The gateway needs to
+            be configured as TCP server and the serial connection to the vallox with 9600 baud, 8-N-1.
+            </description>
```
Please reformat this file using the provided code formatter or clean up settings,

SuperOok replied:
formatted this line manually and then used the Eclipse XML formatter, done

kubawolanin commented on addons/binding/org.openhab.binding.vallox/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ com.google.common.collect;version="10.0.1",
```
Please remove the version constraint

kubawolanin commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,52 @@
+# openhab-vallox
```
Vallox Binding

kubawolanin commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,52 @@
+6. enter configuration parameters (host and port of serial gateway)
+7. The thing should show status `ONLINE`; if not, go into the details of the thing. The status there should include some error message.
+8. make sure you create items for all the many channels of the binding
+  * you can use [demo.items](conf/demo.items) to copy/paste the available items to your own item file
```
I wonder if it gonna work on [auto-generated docs page](http://docs.openhab.org/addons/bindings.html). Maybe it'd be a good idea to include all example Items, Sitemaps, Rules etc. in this file and maybe create a Table of Content?
```
### vallox.things
...

### vallox.items
Number FanSpeed (Vallox) { channel="..." }

### vallox.sitemap
...
```

kubawolanin replied:
Please also include a sample `vallox.things` file

kubawolanin commented on addons/binding/org.openhab.binding.vallox/conf/demo.items:
```diff
@@ -0,0 +1,53 @@
+Group Vallox
+Group ValloxAdvanced
+
+Number FanSpeed (Vallox)
```
Please complete the demo items with corresponding channels, eg:
```
Number FanSpeed (Vallox) { channel="vallox.something.fanSpeed" }
```

kubawolanin commented on addons/binding/org.openhab.binding.vallox/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
Please change `version` to `2.1.0-SNAPSHOT`

kubawolanin commented on addons/binding/org.openhab.binding.vallox/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.vallox</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
```
Please change `version` to `2.1.0-SNAPSHOT`

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,397 @@
+package org.openhab.binding.vallox.handler;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,397 @@
+
+    @Override
+    public void dispose() {
+        logger.info("dispose()");
```
Please change the logging level to `debug`.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,397 @@
+
+    @Override
+    public void initialize() {
+        logger.info("initialize()");
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,397 @@
+                logger.warn("Got update notification for unknown vallox property: {}", channel);
+                break;
+        }
+        // TODO: should check whether value really has changed to reduce amount of updates
```
TODO?

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/ValloxThingHandlerFactory.java:
```diff
@@ -0,0 +1,41 @@
+package org.openhab.binding.vallox.internal;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/StatusChangeListener.java:
```diff
@@ -0,0 +1,16 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,363 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,363 @@
+        byte value = arg;
+
+        if (variable == Variable.IOPORT_FANSPEED_RELAYS.key) {
+            // TODO
```
TODO

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,363 @@
+            vallox.preHeatingOn = (value & 0x10) != 0;
+            vallox.exhaustFanOff = (value & 0x20) != 0;
+            vallox.firePlaceBoosterClosed = (value & 0x40) != 0;
+            // *pUnknown2 = (value & 0x80) != 0;
```
Dead code?

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,363 @@
+            listener.stream().forEach(l -> l.notifyChanged(ValloxProperty.FirePlaceBoosterClosed));
+        } else if (variable == Variable.INSTALLED_CO2_SENSORS.key) {
+            // TODO
+            // convertInstalledCO2Sensor
```
Another TODO :)

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,363 @@
+        } else if (variable == Variable.FIRE_PLACE_BOOSTER_COUNTER.key) {
+            // TODO
+        } else if (variable == Variable.MAINTENANCE_MONTH_COUNTER.key) {
+            // TODO
```
Is this gonna be handled?

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxProperty.java:
```diff
@@ -0,0 +1,102 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxProtocol.java:
```diff
@@ -0,0 +1,45 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+        socket = new Socket(host, port);
+        outputStream = socket.getOutputStream();
+        inputStream = socket.getInputStream();
+        logger.info("Connected to " + host + ":" + port);
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+    }
+
+    public void reconnect() throws UnknownHostException, IOException {
+        logger.info("Trying to reconnect: " + host + ":" + port);
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+        try {
+            if (socket != null) {
+                socket.close();
+                logger.info("Closed " + socket.getInetAddress() + ":" + socket.getPort());
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e1) {
+                        // ignore, simply retry, i.e. wait until reconnected
```
Is this exception gonna be handled e.g. in `trace` logs?

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+                }
+            }
+        });
+        logger.info("Start listening to vallox telegrams!");
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+                }
+            }
+        });
+        logger.info("Start heartbeat check to vallox!");
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+    public void stopListening() {
+        if (listenerExecutor != null) {
+            try {
+                logger.info("attempt to shutdown listener");
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,453 @@
+                    logger.warn("Listener Thread cancel non-finished tasks");
+                }
+                listenerExecutor.shutdownNow();
+                logger.info("shutdown of vallox listener finished");
```
Please change the logging level to `debug`. `trace` can be used for even more details, where necessary.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxStore.java:
```diff
@@ -0,0 +1,223 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValueChangeListener.java:
```diff
@@ -0,0 +1,16 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Variable.java:
```diff
@@ -0,0 +1,281 @@
+package org.openhab.binding.vallox.internal.serial;
```
Every Java file must have a license header. You can run `mvn license:format` on the root of the repo to automatically add missing headers.

kubawolanin commented on addons/binding/pom.xml:
```diff
@@ -60,6 +60,7 @@
   </modules>
 
-</project>
+</project>
```
Please don't delete trailing new line :-)

SuperOok commented:
Hey @kubawolanin , thanks a lot for the many comments. I fixed all of them. However, for some TODOs (e.g. in src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java) there are some telegrams to/from the unit not yet implemented but prepared. These are marked as TODOs for future implementation. I added some more explanation in the method comment.

Still, there seems some issue with the build now.

SuperOok commented:
I'm not a maven pro, but this exception seems to be pretty unrelated...
```
[INFO] Computing target platform for MavenProject: org.openhab.binding:org.openhab.binding.feed:2.1.0-SNAPSHOT @ C:\ws\git\openhab2-addons\addons\binding\org.openhab.binding.feed\pom.xml
[INFO] Resolving dependencies of MavenProject: org.openhab.binding:org.openhab.binding.feed:2.1.0-SNAPSHOT @ C:\ws\git\openhab2-addons\addons\binding\org.openhab.binding.feed\pom.xml
[INFO] {osgi.os=linux, osgi.ws=gtk, org.eclipse.update.install.features=true, osgi.arch=x86}
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: org.openhab.binding.feed 2.1.0.qualifier
[ERROR]   Missing requirement: org.openhab.binding.feed 2.1.0.qualifier requires 'package com.rometools.rome.feed 0.0.0' but it could not be found
```
Any hint?

SuperOok commented:
jääääi, compiles fine. What next?

martinvw commented:
@SuperOok would it be possible to also address my comments? If you have any questions please let us know.

SuperOok commented:
Thanks for all the comments, @martinvw . I saw your very valuable input. Fixing those is on my very long todo-list. Hope to get the chance on addressing them in the next month.

SuperOok commented:
Thanks again for your valuable comments, @martinvw. I reworked my PR to address almost all of them. I reply to you comment if I handle the issue you mentioned differently to your suggestion.
Also I fixed some checkstyle and findbugs findings. Hope it's ok like this for a first try of the binding.

SuperOok commented on addons/binding/org.openhab.binding.vallox/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+Bundle-Name: Vallox Binding
+Bundle-SymbolicName: org.openhab.binding.vallox;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
set to 2.2.0.qualifier now

SuperOok replied:
set to 2.2.0.qualifier

SuperOok commented on addons/binding/org.openhab.binding.vallox/conf/demo.items:
```diff
@@ -0,0 +1,53 @@
+
```
it's in the readme already. removed the seperate files then.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+                logger.warn("Got update notification for unknown vallox property: {}", channel);
+                break;
+        }
+        // XXX: future improvement: check whether value really has changed to reduce amount of updates
```
I removed the XXX due to an forums discussion from which I learn that the optimzation is not necessary because it's already done in the internal channel handling

SuperOok commented on addons/binding/org.openhab.binding.vallox/LICENSE:
```diff
@@ -0,0 +1,203 @@
+Eclipse Public License - v 1.0
```
removed

SuperOok commented on addons/binding/org.openhab.binding.vallox/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+Bundle-SymbolicName: org.openhab.binding.vallox;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
```
ok

SuperOok commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,139 @@
+# Vallox Binding
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/about.html:
```diff
@@ -0,0 +1,28 @@
+and such source code may be obtained at <a href="http://www.openhab.org/">openhab.org</a>.</p>
+
+</body>
+</html>
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/build.properties:
```diff
@@ -0,0 +1,6 @@
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               ESH-INF/
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+                vallox.startHeartbeat();
+                updateStatus(ThingStatus.ONLINE);
+            } catch (IOException e) {
+                String message = "Failed to start connection to Vallox serial interface: " + e.toString();
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+                    Thread.sleep(200);
+                }
+            } else if (command instanceof DecimalType) {
+                byte value = ((DecimalType) command).byteValue();
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+                }
+
+            }
+        } catch (Exception e) {
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+
+            }
+        } catch (Exception e) {
+            logger.error("Failed to handle command to Vallox serial interface: " + e.toString());
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    private static OnOffType getOnOff(boolean on) {
+        if (on) {
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/ValloxThingHandlerFactory.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class ValloxThingHandlerFactory extends BaseThingHandlerFactory {
+
+    private final static Collection<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Lists
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,381 @@
+ */
+public class Telegram {
+
+    private Logger logger = LoggerFactory.getLogger(ValloxSerialHandler.class);
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,381 @@
+
+    private byte sender;
+    private byte receiver;
+    private byte command;
```
Good idea, thanks, makes comparison really more lean, done.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxProperty.java:
```diff
@@ -0,0 +1,109 @@
+ */
+public enum ValloxProperty {
+    // sensor data
+    FanSpeed, // VALLOX_VARIABLE_FAN_SPEED
```
This is one I could not address so far: the binding has quite a number of channels as you might have noticed. They are directly corresponding to this Enum-Literals. In order to not have many many long lists of channel IDs, Enum Literals, Telegram constants and so on, I try to reuse the enum to also denote the channel ID. So I can call ValloxProperty.FanSpeed.name() which will be a valid channel ID also.
I know this is not following naming conventions for enums, but I have not the time to correct this right now just for formal reasons.
I know that the mapping of channels to these Properties to telegrams I receive and send to the physical device is not perfectly done. However, refactoring this would make sense to completely remove that Enum and add some subclassing to telegrams with special telegram classes for each channel and/or telegram handling everything including the channel. However, this is not something I can do at the moment. Hope we can get this initial version despite this issue...

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxProtocol.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class ValloxProtocol {
+
+    public static int BAUDRATE = 9600;
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+
+    public static int MINUTE = 60000;
+
+    private Logger logger = LoggerFactory.getLogger(ValloxSerialHandler.class);
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+    private Socket socket;
+    private ExecutorService listenerExecutor;
+    private ExecutorService heartbeatExecutor;
+    private boolean shutdownListening = false;
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+    }
+
+    public List<ValueChangeListener> getValueListener() {
+        if (vallueListener == null) {
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+
+    public List<StatusChangeListener> getStatusListener() {
+        if (statusListener == null) {
+            statusListener = new ArrayList<StatusChangeListener>();
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+        socket = new Socket(host, port);
+        outputStream = socket.getOutputStream();
+        inputStream = socket.getInputStream();
+        logger.debug("Connected to " + host + ":" + port);
```
ok, done
checked and fixed many others also.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                logger.debug("Closed " + socket.getInetAddress() + ":" + socket.getPort());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+        }
+        Variable v = null;
+        switch (prop) {
+            case AverageEfficiency:
```
Good idea; I will consider using bidirectional maps for mapping channels to telegrams (which is no 1:1 mapping) for the next version of the binding. However, have not time to do that now.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+        telegram[4] = value;
+        telegram[5] = calculateChecksum(telegram);
+
+        logger.debug("sending telegram: {} {} {} {} {} {}", Telegram.byteToHex(telegram[0]),
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+            }
+        } catch (IOException e) {
+            String msg = "Exception writing to vallox: " + e.toString();
+            logger.error(msg);
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                        // telegram too short or invalid
+                        continue;
+                    }
+                    logger.trace(telegram.toString());
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    logger.trace(telegram.toString());
+                    telegram.updateStore(vallox, vallueListener);
+                } catch (Exception e) {
+                    logger.error("Exception reading input stream: " + e.toString());
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    logger.error("Exception reading input stream: " + e.toString());
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e1) {
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+     * automatically reconnects if it fails.
+     */
+    public void startHeartbeat() {
+        heartbeatExecutor = Executors.newSingleThreadExecutor();
```
Good idea, makes the code much more clean, done.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    sendPoll(ValloxProperty.SelectStatus);
+                } catch (Exception e) {
+                    String msg = "Exception sending heartbeat poll: " + e.toString();
+                    logger.error(msg);
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    try {
+                        this.reconnect();
+                    } catch (IOException e1) {
+                        logger.error("Exception reconnecting vallox: " + e.toString());
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+        int available = inputStream.available();
+        if (available < ValloxProtocol.LENGTH) {
+            Thread.sleep(200);
+            return null;
```
Good idea, I reworked that and introduced some exceptions, done.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                shutdownListening = true;
+                listenerExecutor.shutdown();
+                listenerExecutor.awaitTermination(5, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxStore.java:
```diff
@@ -0,0 +1,230 @@
+public class ValloxStore {
+
+    // sensor data
+    public int fanSpeed; // VALLOX_VARIABLE_FAN_SPEED
```
Yah, it's not optimal, quite simple solution, I know. As stated above I might rework that in a second version of the binding totally.

SuperOok commented on addons/binding/pom.xml:
```diff
@@ -72,6 +72,7 @@
     <module>org.openhab.binding.tesla</module>
     <module>org.openhab.binding.toon</module>
     <module>org.openhab.binding.urtsi</module>
+    <module>org.openhab.binding.vallox</module>
```
ok, done

SuperOok commented:
@kaikreuzer , the first three commits unfortunately lack the signed-off-by. How can I fix that? I'm not a git pro and I found only some hints in the web about fixing them for unpushed commits. However, these are already pushed; so what can I do? Any hint?

martinvw commented:
Thanks for your changes.

Regarding the changing of commits, the explanation I always provide:

>Can you squash your commits and change commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work
>
>For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Does it help, any questions just ask.

SuperOok commented:
Ok, I did squashing. Hope I did it right...

SuperOok commented:
After detailed testing of configuration I added a few small fixes to the pull request. LGTM now. How to proceed?

martinvw commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxStore.java:
```diff
@@ -0,0 +1,231 @@
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
```
The beauty of the StringBuilder is that you can chain it. Next time please generate it, safes you some work.

SuperOok replied:
I already used the Eclipse toString() generator. It has two styles to generate with the StringBuilder, either with or without chaining. I now changed it to chaining. It's a bit shorter.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,421 @@
+                this.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, message + e.getMessage());
+            } catch (NumberFormatException e) {
+                String message = "Failed to read input for port parameter of Vallox interface. Must be a plain positive integer.";
+                logger.error(message, e);
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/build.properties:
```diff
@@ -0,0 +1,8 @@
+               .,\
+               OSGI-INF/,\
+               ESH-INF/,\
+               README.md,\
```
ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,460 @@
+
+        byte value = arg;
+
+        switch (commandVar) {
```
Oops, that was a big bug. I wonder how that have passed my regression test. :-| Thanks for the hint. Fixed.

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/Telegram.java:
```diff
@@ -0,0 +1,460 @@
+            }
+            case BASIC_HUMIDITY_LEVEL: {
+                vallox.basicHumidityLevel = Byte.toUnsignedInt(value);
+                listener.stream().forEach(l -> l.notifyChanged(ValloxProperty.BasicHumidityLevel));
```
Ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,473 @@
+
+    public static final int MINUTE = 60000;
+
+    private Logger logger = LoggerFactory.getLogger(ValloxSerialInterface.class);
```
Ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,473 @@
+    private ScheduledExecutorService heartbeatExecutor;
+    private boolean shutdownListening;
+
+    private List<ValueChangeListener> valueListener;
```
Ok, done

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,473 @@
+            }
+        } catch (IOException e) {
+            String msg = "Exception writing to vallox.";
+            logger.error(msg, e);
```
Ok, done

kaikreuzer replied:
so please reduce to debug level

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,473 @@
+     * automatically reconnects if it fails.
+     */
+    public void startHeartbeat() {
+        heartbeatExecutor = Executors.newScheduledThreadPool(1);
```
Ok, I pass the scheduler of the BaseThingHandler now to the downstream classes and use it there. Good hint, thanks!

SuperOok commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,473 @@
+                }
+            }
+        };
+        heartbeatExecutor.scheduleAtFixedRate(reconnecter, 1, 1, TimeUnit.MINUTES);
```
Ok, done

SuperOok commented:
Thanks again, @martinvw, for your deep review. You really give good hints to make the code clean!

sjka commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,421 @@
+                    port = (BigDecimal) portObject;
+                }
+                // this is the case when parameter is set through .thing-File
+                else if (portObject instanceof String) {
```
AFAIK it is, I'm not aware that there is anything open still. 
Please rather open an issue in https://github.com/eclipse/smarthome with steps to reproduce than working around it in the binding.

martinvw replied:
@SuperOok can you please take a look at this.

martinvw commented:
O o I see a lot of follow up PR's, it would be great if you could remove the workaround I will also queue for a review by @kaikreuzer the owner of the project. Its great to see that your contribution has already a lot of incoming Pr's too.

SuperOok commented:
OK, I fixed the workaround. Maybe the original problem was that the configuration in the thing file was `port="26"` as I was not aware that you can use numbers directly as `port=26`. So my fault. Works now for me.

Hilbrand commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -369,6 +369,11 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.toon/${project.version}</bundle>
     </feature>
+	
+	<feature name="openhab-binding-vallox" description="Vallox Binding" version="${project.version}">
```
indentation: tab must be spaces here. Also remove tab in line above.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+				    The gateway needs to be configured as TCP server and the serial 
+				    connection to the vallox with 9600 baud, 8-N-1.
+                </description>
+				<default>lwip</default>
```
Could you re-apply the code formatter? There seems to be a tabs/spaces mix.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+                </description>
+				<default>lwip</default>
+			</parameter>
+			<parameter name="port" type="integer" required="true">
```
Not required as you have a default

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="FanSpeedMax">
+		<item-type>Number</item-type>
+		<label>FanSpeedMax</label>
```
Better "Max. Fan Speed"?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="FanSpeedMin">
+		<item-type>Number</item-type>
+		<label>FanSpeedMin</label>
```
Better "Min. Fan Speed"?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="TempInside">
+		<item-type>Number</item-type>
+		<label>Temp Inside</label>
```
Better: Inside Temperature

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="TempOutside">
+		<item-type>Number</item-type>
+		<label>Temp Outside</label>
```
Better: Outside Temperature

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="TempExhaust">
+		<item-type>Number</item-type>
+		<label>Temp Exhaust</label>
```
Better: Exhaust Temperature

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="TempIncomming">
+		<item-type>Number</item-type>
+		<label>Temp Incomming</label>
```
Better: Incoming Temperature

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="InEfficiency">
+		<item-type>Number</item-type>
+		<label>InEfficiency</label>
```
What is this? Anyhow, don't use CamelCase in labels...

gitMiguel replied:
All these efficiencies are percent calculations from different ventilation system temperatures. Per manufacturer description.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="OutEfficiency">
+		<item-type>Number</item-type>
+		<label>OutEfficiency</label>
```
What is this? Anyhow, don't use CamelCase in labels...

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="AverageEfficiency">
+		<item-type>Number</item-type>
+		<label>AverageEfficiency</label>
```
What is this? Anyhow, don't use CamelCase in labels...

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="CO2AdjustState">
+		<item-type>Switch</item-type>
+		<label>CO2 Adjust State</label>
```
Better: CO2 Adjustment?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="HumidityAdjustState">
+		<item-type>Switch</item-type>
+		<label>Humidity Adjust State</label>
```
Better: Humidity Adjustment?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="DCFanInputAdjustment" advanced="true">
+		<item-type>Number</item-type>
+		<label>DCFanInputAdjustment</label>
```
all following: Please no CamelCase for labels

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+		<item-type>Number</item-type>
+		<label>DCFanInputAdjustment</label>
+		<description>0-100%</description>
+		<state min="0" max="255" step="1" pattern="%d" readOnly="false"></state>
```
0-255 contradicts 0-100% from the description

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+		<item-type>Number</item-type>
+		<label>DCFanOutputAdjustment</label>
+		<description>0-100%</description>
+		<state min="0" max="255" step="1" pattern="%d" readOnly="false"></state>
```
dito

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+	<channel-type id="LastErrorNumber" advanced="true">
+		<item-type>Number</item-type>
+		<label>LastErrorNumber</label>
+		<description>  5: Supply air temperature sensor fault; 6: Carbon dioxide alarm; 7: Outdoor air sensor fault;
```
Note that you can use HTML elements and CDATA sections in the descriptions - this looks as if you'd want to put it in a bullet point list.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+Bundle-Name: Vallox Binding
+Bundle-SymbolicName: org.openhab.binding.vallox;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
```
Sorry for being late. Might I ask you to update this to 2.3.0.qualifier?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+
+## Prerequisite
+
+This Openhab2 bundle so far only works with a RS485-to-ethernet gateway.
```
-> This binding...

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+## Prerequisite
+
+This Openhab2 bundle so far only works with a RS485-to-ethernet gateway.
+It does not support a local serial interface of the device executing OH2.
```
remove "of the device executing OH2."

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+This Openhab2 bundle so far only works with a RS485-to-ethernet gateway.
+It does not support a local serial interface of the device executing OH2.
+
+I do not have any experience with direct serial RS485 communication under Java and
```
remove this statement here - rather find people in the forum.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+* parity: none
+* stop bits: 1 
+
+## Install
```
Please remove this section - this is imho nothing special (and should not be special) for this binding, but the normal way of configuring the system.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+
+## Features
+
+The binding currently supports the following Features
```
-> features

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+* 14 major properties, some read-only, some writeable
+* 35 advanced properties, so far read-only most of them
+
+Note that there might be more properties that can be read or written which might be not implemented yet. Contributions welcome.
```
Remove this, this is not really end user compatible. Also don't do links to the code here, they will break on the official site.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+## Features
+
+The binding currently supports the following Features
+* 14 major properties, some read-only, some writeable
```
Properties? Wouldn't this be the right place to list the available channels and there ids?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+Note that there might be more properties that can be read or written which might be not implemented yet. Contributions welcome.
+Also not all channels are perfectly documented. See [thing-types.xml](ESH-INF/thing/thing-types.xml) for details.
+
+* the binding will automatically detect loss of connection (at latest by a heartbeat poll to the vallox every minute)
```
Can imho also be removed - this is rather expected behavior...

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/README.md:
```diff
@@ -0,0 +1,143 @@
+### vallox.things
+
+It's easiest to define the thing textually and map the items textually to the thing's channels.
+Parameters are
```
Those should be mentioned in a "Thing configuration" section above, not only here under "Examples".
Check the official template for the binding README.

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/pom.xml:
```diff
@@ -0,0 +1,20 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.vallox</artifactId>
+  <version>2.2.0-SNAPSHOT</version>
```
remove this line, the version is inherited from the parent

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/pom.xml:
```diff
@@ -0,0 +1,20 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.2.0-SNAPSHOT</version>
```
2.3.0-SNAPSHOT

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,414 @@
+                updateStatus(ThingStatus.ONLINE);
+            } catch (IOException e) {
+                String message = "Failed to start connection to Vallox serial interface. ";
+                logger.warn(message, e);
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/handler/ValloxSerialHandler.java:
```diff
@@ -0,0 +1,414 @@
+                this.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, message + e.getMessage());
+            } catch (NumberFormatException e) {
+                String message = "Failed to read input for port parameter of Vallox interface. Must be a plain positive integer. Don't use quotes in a .thing-File!";
+                logger.warn(message, e);
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+    public void close() {
+        try {
+            if (socket != null) {
+                socket.close();
```
Are the input and output streams implicitly closed as well here?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    logger.trace("Received message: {}", telegram);
+                    telegram.updateStore(vallox, valueListener);
+                } catch (MalformedTelegramException e) {
+                    logger.warn("Issue receiving telegram. Discarding.", e);
```
Is this a bug situation that users should report as an issue if they see this?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                    // telegram too short, wait for more data, but ignore
+                    sleep = 200;
+                } catch (IOException e) {
+                    logger.error("Exception reading input stream.", e);
```
Is this a bug situation that users should report as an issue if they see this?
Can't we also get here if the connection is simply lost?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                            Thread.sleep(sleep);
+                            sleep = 0; // reset sleep for the next telegram
+                        } catch (InterruptedException e1) {
+                            logger.error("Telegram listening thread interrupted.");
```
Is this a bug situation that users should report as an issue if they see this or could it happen during a regular system shutdown?

kaikreuzer commented on addons/binding/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/serial/ValloxSerialInterface.java:
```diff
@@ -0,0 +1,465 @@
+                        try {
+                            vallox.reconnect();
+                        } catch (IOException e1) {
+                            logger.error("Exception reconnecting vallox.", e1);
```
Is this a bug situation that users should report as an issue if they see this?

gitMiguel commented:
Hi,

Not able to do any reviews but I'll leave few comments here.

I couldn't get this binding to run so I cloned SuperOok's repo and made some changes foe my personal use. Firstly I had problems with ScheduledExecutor as it almost all the time got rejected exception and of course wouldn't run even after multiple restarts. I created two new executors one for listener and one for heartBeat. Might be a workaround but works for me. Second problem was with line 434 in ValloxSerialInterface. MessageFormat.format() didn't work as expected and would block indefinetly as soon as a telegram that wasn't meant for us arrived. I changed MessageFormat.format() to String.format() and used %d as placeholders in msg.

Sprocket02 commented:
@gitMiguel: I would Like to test your binding. Can you provide me a binary?

gitMiguel commented:
Find my repo here: [https://github.com/gitMiguel/openhab2-addons/tree/%231559-vallox-binding-latest/addons/binding/org.openhab.binding.vallox](url) There's also a build included. Just rename .zip jo .jar. It's not the latest version but it's a start for testing. I'm in the middle of implementing remaining channels and write commands. I have enabled issues in my repo so let's continue this possible discussion there. Watch my repo to get informed about new versions.

/Miika

majherek commented:
Hi @gitMiguel,

in free time I will also test yours binding. I use openhab 2.2 with org.eclipse.smarthome.binding.vallox-2.0.0-SNAPSHOT.jar which I compile from SuperOok repo. But I am afraid that this plugin won't work on OH2 2.3.

So I also will try to test it ASAP.

kaikreuzer commented:
@SuperOok Are you still around and plan to incorporate the review comments and update this PR?

Sprocket02 commented:
Hi,

I see the binding in the Paper UI, but there is a problem with parsing the vallox.things file (openhab.log):

`2018-04-13 20:21:56.485 [WARN ] [el.core.internal.ModelRepositoryImpl] - Configuration model 'vallox.things' has errors, therefore ignoring it: [1,7]: no viable alternative at input 'vallox'
[1,14]: no viable alternative at input 'serial'
[1,21]: no viable alternative at input 'main'
[1,28]: no viable alternative at input 'serialPort'
[1,55]: no viable alternative at input 'panelNumber'
[1,67]: no viable alternative at input '5'`

My vallox things file:

`Thing vallox.serial.main [ serialPort="/dev/ttyUSB0", panelNumber=1 ]`

In the Visual Studio Editor, I get an error message in the vallox.things file:

`no viable alternative at input 'vallox'
no viable alternative at input 'serial'
no viable alternative at input 'main'
no viable alternative at input 'serialPort'
no viable alternative at input 'panelNumber'
no viable alternative at input '1'`

Has anybody an idea?

Sprocket02 commented:
Oh I'm using SNAPSHOT 2.3.0

SuperOok commented:
@kaikreuzer: Sorry that it takes so long. Yes, in general I'm still around and willing to finalize the PR. However, as the PR started quite long ago, life moves on and unfortunately I lost my concrete system-under-test in the meantime if you know what I mean. Hence, personally I have very limited testing capabilities at the moment. But I can address the comments and make it run and hope someone else can do the final test on a real Vallox installation.
Hope to get my fingers at it till summer. Cheers.

martinvw commented:
@SuperOok any update from you? Thanks for your time!

davidgraeff commented:
Closing as the original contributor doesn't have the equipment anymore
